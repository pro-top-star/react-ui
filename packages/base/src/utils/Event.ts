import { ReactNode } from 'react';
import { HTMLEvent } from '../interfaces/HTMLEvent';

export interface Parameters {
  [key: string]: number | string | object | boolean | Parameters;
}

export class Event {
  readonly source: ReactNode;
  readonly originalEvent: HTMLEvent;
  readonly parameters: any;
  readonly htmlSource: EventTarget;

  /**
   * Event Constructor
   * @param source {ReactNode} The React Element instance
   * @param originalEvent {HTMLEvent} original HTML Event that triggered this event
   * @param parameters {Parameters} event payload
   */
  constructor(source: ReactNode, originalEvent: HTMLEvent, parameters: Parameters = {}) {
    this.source = source;
    this.originalEvent = originalEvent;
    this.parameters = parameters;
    this.htmlSource = originalEvent.target;
  }

  /**
   * Static initializer for an Event
   * @param source {ReactNode} The React Element instance
   * @param originalEvent {HTMLEvent} original HTML Event that triggered this event
   * @param parameters {Parameters} event payload
   */
  public static of(source: ReactNode, originalEvent: HTMLEvent, parameters: Parameters = {}) {
    return new Event(source, originalEvent, parameters);
  }

  public getSource(): ReactNode {
    return this.source;
  }

  public getOriginalEvent(): HTMLEvent {
    return this.originalEvent;
  }

  public getParameters(): Parameters {
    return this.parameters;
  }

  public getParameter(parameter: string): any {
    return this.parameters[parameter];
  }

  public getHtmlSourceElement(): EventTarget {
    return this.htmlSource;
  }

  public cancel(): void {
    this.originalEvent.stopPropagation();
  }

  public preventDefault(): void {
    this.originalEvent.preventDefault();
  }
}
