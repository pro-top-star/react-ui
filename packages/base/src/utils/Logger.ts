/* eslint-disable no-console,no-unused-expressions */

const pad0 = (i: any, w: number) => `000${String(i)}`.slice(-w);

export enum LOG_LEVEL {
  FATAL = 0,
  ERROR = 1,
  WARNING = 2,
  INFO = 3,
  DEBUG = 4,
  TRACE = 5
}

class DeviceLogger {
  public defaultComponent: string;
  public sWindowName = window.top === window ? '' : `[${window.location.pathname.split('/').slice(-1)[0]}] `;

  constructor(defaultComponent = 'DEVICE') {
    this.log(LOG_LEVEL.INFO, 'Device API logging initialized');
    this.defaultComponent = defaultComponent;
    return this;
  }

  public for(defaultComponent: string) {
    return new DeviceLogger(defaultComponent);
  }

  public log(iLevel: LOG_LEVEL, sMessage: string, sComponent?: string) {
    sComponent = sComponent || this.defaultComponent || '';
    const oNow = new Date();
    const oLogEntry = {
      time: `${pad0(oNow.getHours(), 2)}:${pad0(oNow.getMinutes(), 2)}:${pad0(oNow.getSeconds(), 2)}`,
      date: `${pad0(oNow.getFullYear(), 4)}-${pad0(oNow.getMonth() + 1, 2)}-${pad0(oNow.getDate(), 2)}`,
      timestamp: oNow.getTime(),
      level: iLevel,
      message: sMessage || '',
      component: sComponent || ''
    };
    if (window.console) {
      const logText = `${oLogEntry.date} ${oLogEntry.time} ${this.sWindowName}${oLogEntry.message} - ${oLogEntry.component}`;
      switch (iLevel) {
        case LOG_LEVEL.FATAL:
        case LOG_LEVEL.ERROR:
          console.error(logText);
          break;
        case LOG_LEVEL.WARNING:
          console.warn(logText);
          break;
        case LOG_LEVEL.INFO:
          console.info ? console.info(logText) : console.log(logText);
          break; // info not available in iOS simulator
        case LOG_LEVEL.DEBUG:
          console.debug ? console.debug(logText) : console.log(logText);
          break; // debug not available in IE, fallback to log
        case LOG_LEVEL.TRACE:
          console.trace ? console.trace(logText) : console.log(logText);
          break; // trace not available in IE, fallback to log
      }
    }
    return oLogEntry;
  }

  //  Public API
  public fatal(msg: string, component?: string) {
    return this.log(LOG_LEVEL.FATAL, msg, component);
  }

  public error(msg: string, component?: string) {
    return this.log(LOG_LEVEL.ERROR, msg, component);
  }

  public warning(msg: string, component?: string) {
    return this.log(LOG_LEVEL.WARNING, msg, component);
  }

  public info(msg: string, component?: string) {
    return this.log(LOG_LEVEL.INFO, msg, component);
  }

  public debug(msg: string, component?: string) {
    return this.log(LOG_LEVEL.DEBUG, msg, component);
  }

  public trace(msg: string, component?: string) {
    return this.log(LOG_LEVEL.TRACE, msg, component);
  }
}

export const Logger = new DeviceLogger();
