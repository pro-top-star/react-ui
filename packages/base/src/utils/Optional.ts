/* eslint-disable id-blacklist */
/**
 * Optionals for Javascript Objects
 * Inspired by Java Optionals
 */
class Optional {
  private readonly object: object;
  private path: string;

  /**
   * Creates a new Optional
   * @param {Object} object The Object to validate
   * @param {String} path Path to search for in object. Separated by dots.
   *                      Example: 'foo.bar'
   * @constructor
   */
  constructor(object: object, path: string) {
    this.object = object;
    this.path = path;
  }

  /**
   * Creates a new Optional
   * @param {Object} object The Object to validate
   * @param {String} path Path to search for in object. Separated by dots.
   *                      Example: 'foo.bar'
   * @constructor
   */
  static of(object: object, path: string) {
    return new Optional(object, path);
  }

  /**
   * Get The Value of the Optional.
   * Do not use the this method directly!
   * @return {any} Resolved Value
   * @private
   */
  get() {
    return this.path.split('.').reduce((acc, val) => {
      if (acc === undefined || acc === null) {
        return acc;
      }
      if (acc.hasOwnProperty(val)) {
        const newVal = acc[val];
        if (newVal !== undefined && newVal !== null) {
          return newVal;
        }
        return newVal;
      }
      return undefined;
    }, this.object);
  }

  /**
   * Return true if there is a value present, otherwise false.
   * @returns {Boolean} true if there is a value present, otherwise false
   */
  isPresent() {
    let obj: any = this.object;
    return this.path.split('.').every((x) => {
      if (typeof obj !== 'object' || obj === null || obj === undefined || !(x in obj)) return false;
      obj = obj[x];
      return true;
    });
  }

  /**
   * If a value is present, invoke the specified consumer with the value, otherwise do nothing.
   * @param {Function} fn block to be executed if a value is present
   * @returns {any} value The executed fn if present, else null
   */
  ifPresent(fn: (value: any) => void) {
    if (this.isPresent()) {
      return fn(this.get());
    }
    return null;
  }

  /**
   * Return the value if present, otherwise return other.
   * @param val {any} the value to be returned if there is no value present, may be null
   * @returns {any} value the value, if present, otherwise other
   */
  orElse(val: any) {
    return this.isPresent() ? this.get() : val;
  }

  /**
   * Return the contained value, if present, otherwise throw an exception to be created by the provided supplier.
   * @param {ExceptionSupplier} exceptionSupplier The supplier which will return the exception to be thrown
   * @returns {any} the present value
   * @throws {Exception} ex - if there is no value present
   *
   */
  orElseThrow(exceptionSupplier: () => Error) {
    if (!this.isPresent()) {
      throw exceptionSupplier();
    }
    return this.get();
  }
}

export { Optional };
