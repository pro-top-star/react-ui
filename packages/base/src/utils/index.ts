import { UIEvent } from 'react';

export const deprecationNotice = (component: string, message: string) => {
  if (process.env.NODE_ENV === 'development') {
    const value = `*** ui5-webcomponents-react Deprecation Notice - ${component} ***\n`;
    // eslint-disable-next-line no-console
    console.warn(`${value}${message}`);
  }
};

export const getScrollBarWidth = () => {
  const inner = document.createElement('p');
  inner.style.width = '100%';
  inner.style.height = '200px';

  const outer = document.createElement('div');
  outer.style.position = 'absolute';
  outer.style.top = '0px';
  outer.style.left = '0px';
  outer.style.visibility = 'hidden';
  outer.style.width = '200px';
  outer.style.height = '150px';
  outer.style.overflow = 'hidden';
  outer.appendChild(inner);

  document.body.appendChild(outer);
  const w1 = inner.offsetWidth;
  outer.style.overflow = 'scroll';
  let w2 = inner.offsetWidth;

  if (w1 === w2) {
    w2 = outer.clientWidth;
  }

  document.body.removeChild(outer);
  return w1 - w2;
};

export const enrichEventWithDetails = <T = {}>(event: UIEvent, payload: T = {} as any) => {
  if (event.hasOwnProperty('persist')) {
    // if there is a persist method, it's an SyntheticEvent so we need to persist it
    event.persist();
  }

  const shouldCreateNewDetails =
    event.detail === null || event.detail === undefined || typeof event.detail !== 'object';
  Object.defineProperty(event, 'detail', {
    value: shouldCreateNewDetails ? {} : event.detail,
    writable: true,
    configurable: true
  });
  Object.assign(event.detail, payload);
  return (event as unknown) as CustomEvent<T>;
};
