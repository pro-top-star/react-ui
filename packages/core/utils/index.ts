export * from './Device';
export * from './isNumeric';
