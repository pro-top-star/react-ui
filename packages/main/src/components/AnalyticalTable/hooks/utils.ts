export const makeTemplateColumns = (columns, resizedColumns) => {
  return columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
};
