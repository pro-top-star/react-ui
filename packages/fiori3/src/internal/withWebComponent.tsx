import { Event } from '@ui5-webcomponents-react/utils';
import React, { Children, cloneElement, Component, CSSProperties, FC, ReactElement } from 'react';
import { withTheme } from 'react-jss';
import { Fiori3CommonProps } from '../interfaces/Fiori3CommonProps';
import { JSSTheme } from '../interfaces/JSSTheme';
import { Ui5DomRef } from '../interfaces/Ui5DomRef';
import { Ui5WebComponentMetadata } from '../interfaces/Ui5WebComponentMetadata';

function capitalizeFirstLetter(s: string) {
  return s.charAt(0).toUpperCase() + s.slice(1);
}

function toKebabCase(s: string) {
  return s.replace(/([A-Z])/g, (a, b) => `-${b.toLowerCase()}`);
}

export interface WithWebComponentPropTypes extends Fiori3CommonProps {
  innerComponentRef?: (el: HTMLElement) => void;
  children?: any | void;
  innerStyles?: CSSProperties;
}

interface WithWebComponentInternalPropTypes extends WithWebComponentPropTypes {
  theme: JSSTheme;
}

export function withWebComponent<T>(WebComponent): FC<T> {
  class WithWebComponent extends Component<T & WithWebComponentPropTypes> {
    state = {
      updateAfterMount: false
    };

    private wcRef: Ui5DomRef;
    private shadowRootRef: HTMLElement;
    private eventRegistry = {};
    private eventRegistryWrapped = {};

    private handleInnerRef = (ref: HTMLElement) => {
      const { innerComponentRef } = this.props;
      this.wcRef = ref;
      if (innerComponentRef) {
        innerComponentRef(ref);
      }
    };

    private getWebComponentMetadata = (): Ui5WebComponentMetadata => {
      if (WebComponent) {
        return WebComponent.getMetadata();
      }

      return {
        getProperties() {
          return {};
        },
        getEvents() {
          return {};
        },
        getSlots() {
          return {};
        }
      };
    };

    private getMetadataBooleans = () => {
      return Object.entries(this.getWebComponentMetadata().getProperties())
        .filter(([key, value]) => value.type === Boolean)
        .map(([key]) => key);
    };

    private createEventWrapperFor = (eventIdentifier, eventHandler) => (e) => {
      let payload = Object.keys(this.getWebComponentMetadata().getProperties()).reduce((acc, val) => {
        if (/^_/.test(val)) {
          return acc;
        }
        acc[val] = e.target[toKebabCase(val)];
        return acc;
      }, {});

      payload = Object.keys(this.getWebComponentMetadata().getEvents()[eventIdentifier]).reduce((acc, val) => {
        if (val === 'detail' && e[val]) {
          return {
            ...acc,
            ...e[val]
          };
        }
        acc[val] = (e.detail && e.detail[val]) || e[val];
        return acc;
      }, payload);
      eventHandler(Event.of(this, e, payload));
    };

    private bindEvents = () => {
      this.getMetadataEvents().forEach((eventIdentifier) => {
        const alternativeKey = 'on' + capitalizeFirstLetter(eventIdentifier);
        const eventHandler = this.props[eventIdentifier] || this.props[alternativeKey];
        if (typeof eventHandler === 'function' && this.eventRegistry[alternativeKey] !== eventHandler) {
          if (this.eventRegistry[alternativeKey]) {
            this.wcRef.removeEventListener(eventIdentifier, this.eventRegistryWrapped[alternativeKey]);
          }
          this.eventRegistryWrapped[alternativeKey] = this.createEventWrapperFor(eventIdentifier, eventHandler);
          this.wcRef.addEventListener(eventIdentifier, this.eventRegistryWrapped[alternativeKey]);
          this.eventRegistry[alternativeKey] = eventHandler;
        } else if (this.eventRegistry[alternativeKey] && !eventHandler) {
          this.wcRef.removeEventListener(eventIdentifier, this.eventRegistryWrapped[alternativeKey]);
        }
      });
    };

    private getMetadataSlots = () => {
      return Object.keys(this.getWebComponentMetadata().getSlots());
    };

    private getMetadataEvents = () => {
      return Object.keys(this.getWebComponentMetadata().getEvents());
    };

    get booleanProps() {
      return this.getMetadataBooleans().reduce((acc, key) => {
        if (this.props[key]) {
          acc[toKebabCase(key)] = true;
        }
        return acc;
      }, {});
    }

    static get propBlacklist() {
      return {
        className: true,
        innerStyles: true
      };
    }

    get shadowDomRef() {
      if (this.shadowRootRef) {
        return this.shadowRootRef;
      }
      return (this.shadowRootRef = this.wcRef && this.wcRef.getDomRef ? this.wcRef.getDomRef() : null);
    }

    get regularProps() {
      if (this.wcRef) {
        this.bindEvents();
      }

      const props = {};
      const slotProps = {};

      Object.entries(this.props)
        .filter(([key]) => !this.getMetadataBooleans().includes(key))
        .filter(
          ([key]) =>
            !this.getMetadataEvents().some(
              (eventKey) => `on${capitalizeFirstLetter(eventKey)}` === key || key === eventKey
            )
        )
        .filter(([key]) => !WithWebComponent.propBlacklist[key])
        .forEach(([key, value]) => {
          if (this.getMetadataSlots().includes(key)) {
            slotProps[key] = value;
          } else {
            props[toKebabCase(key)] = value;
          }
        });

      return { props, slotProps };
    }

    applyInnerStyles = () => {
      const { innerStyles } = this.props;
      const shadowRef = this.shadowDomRef;
      if (!shadowRef) {
        return;
      }
      if (innerStyles) {
        Object.entries(innerStyles).forEach(([key, value]) => {
          shadowRef.style[key] = value;
        });
      }
    };

    removeOldStyles = (prevStyles) => {
      if (prevStyles) {
        Object.keys(prevStyles).forEach((key) => {
          this.shadowDomRef.style[key] = '';
        });
      }
    };

    componentDidUpdate(prevProps) {
      const executionList = [];

      if (prevProps.innerStyles !== this.props.innerStyles) {
        executionList.push(() => this.removeOldStyles(prevProps.innerStyles));
        executionList.push(this.applyInnerStyles);
      }

      if (executionList.length > 0) {
        requestAnimationFrame(() => {
          executionList.forEach((exec) => exec());
        });
      }
    }

    componentDidMount() {
      if (this.wcRef) {
        this.bindEvents();
        requestAnimationFrame(() => {
          this.applyInnerStyles();
        });
      } else {
        this.setState({
          updateAfterMount: true
        });
      }
    }

    render() {
      const { className = '' } = this.props;

      const CustomTag = WebComponent.getMetadata().getTag();
      const slots = WebComponent.getMetadata().getSlots();

      const { props, slotProps } = this.regularProps;
      const { children, theme, ...regularProps } = props as T & WithWebComponentInternalPropTypes;
      return (
        <CustomTag {...this.booleanProps} ref={this.handleInnerRef} {...regularProps} class={className}>
          {Object.keys(slots).map((slot) => {
            if (slotProps[slot]) {
              return Children.map(slotProps[slot], (item: ReactElement<any>, index) =>
                cloneElement(item, {
                  key: `${slot}-${index}`,
                  'data-ui5-slot': slot
                })
              );
            }
            return null;
          })}
          {children}
        </CustomTag>
      );
    }
  }

  return withTheme(WithWebComponent);
}
