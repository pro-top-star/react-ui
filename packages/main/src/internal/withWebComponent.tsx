import { Event } from '@ui5/webcomponents-react-base/lib/Event';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import React, {
  Children,
  cloneElement,
  ReactElement,
  Ref,
  RefForwardingComponent,
  RefObject,
  useEffect,
  useRef
} from 'react';
import { CommonProps } from '../interfaces/CommonProps';
import { Ui5DomRef } from '../interfaces/Ui5DomRef';
import { Ui5WebComponentMetadata } from '../interfaces/Ui5WebComponentMetadata';

function capitalizeFirstLetter(s: string) {
  return s.charAt(0).toUpperCase() + s.slice(1);
}

function toKebabCase(s: string) {
  return s.replace(/([A-Z])/g, (a, b) => `-${b.toLowerCase()}`);
}

const REACT_SPECIFIC_ATTRIBUTES = ['className'];

export interface WithWebComponentPropTypes extends CommonProps {
  ref?: Ref<any>;
  children?: any | void;
}

export function withWebComponent<T>(WebComponent): RefForwardingComponent<Ui5DomRef, T & WithWebComponentPropTypes> {
  const getWebComponentMetadata = (): Ui5WebComponentMetadata => {
    if (WebComponent) {
      return WebComponent.getMetadata();
    }

    return {
      metadata: {
        events: {}
      },
      getProperties() {
        return {};
      },
      getSlots() {
        return {};
      },
      getEvents() {
        return {};
      }
    };
  };

  const getBooleanPropsFromMetadata = (): Array<string> => {
    return Object.entries(getWebComponentMetadata().getProperties())
      .filter(([key, value]) => value.type === Boolean)
      .map(([key]) => key);
  };

  const getSlotsFromMetadata = (): Array<string> => {
    return Object.keys(getWebComponentMetadata().getSlots());
  };

  const getEventsFromMetadata = (): Array<string> => {
    return Object.keys(getWebComponentMetadata().getEvents() || {}).filter((eventName) => !eventName.startsWith('_'));
  };

  const createEventWrapperFor = (eventIdentifier, eventHandler) => (e) => {
    let payload = Object.keys(getWebComponentMetadata().getProperties()).reduce((acc, val) => {
      if (/^_/.test(val)) {
        return acc;
      }
      acc[val] = e.target[toKebabCase(val)];
      return acc;
    }, {});

    const eventMeta = getWebComponentMetadata().getEvents()[eventIdentifier] || {};

    payload = Object.keys(eventMeta).reduce((acc, val) => {
      if (val === 'detail' && e[val]) {
        return {
          ...acc,
          ...e[val]
        };
      }
      acc[val] = (e.detail && e.detail[val]) || e[val];
      return acc;
    }, payload);
    // TODO: Pass Web Component Ref in here?
    eventHandler(Event.of(null, e, payload));
  };

  const WithWebComponent = React.forwardRef((props: T & WithWebComponentPropTypes, wcRef: RefObject<Ui5DomRef>) => {
    const eventRegistry = useRef({});
    const eventRegistryWrapped = useRef({});

    const CustomTag = WebComponent.getMetadata().getTag();
    const ref = useConsolidatedRef<HTMLDivElement>(wcRef);

    const getBooleanProps = () => {
      return getBooleanPropsFromMetadata().reduce((acc, key) => {
        if (props[key]) {
          acc[toKebabCase(key)] = true;
        }
        return acc;
      }, {});
    };

    useEffect(
      () => {
        getEventsFromMetadata().forEach((eventName) => {
          const eventPropName = 'on' + capitalizeFirstLetter(eventName);
          const eventHandler = props[eventPropName];

          if (typeof eventHandler === 'function' && eventRegistry.current[eventPropName] !== eventHandler) {
            if (eventRegistry.current[eventPropName]) {
              ref.current.removeEventListener(eventName, eventRegistryWrapped.current[eventPropName]);
            }
            eventRegistryWrapped.current[eventPropName] = createEventWrapperFor(eventName, eventHandler);
            ref.current.addEventListener(eventName, eventRegistryWrapped.current[eventPropName]);
            eventRegistry.current[eventPropName] = eventHandler;
          } else if (eventRegistry.current[eventPropName] && !eventHandler) {
            ref.current.removeEventListener(eventName, eventRegistryWrapped.current[eventPropName]);
          }
        });
      },
      getEventsFromMetadata().map((eventName) => props[`on${capitalizeFirstLetter(eventName)}`])
    );

    const { className, ...otherProps } = props;

    const getRegularProps = () => {
      return Object.entries(otherProps)
        .filter(([key]) => !getBooleanPropsFromMetadata().includes(key))
        .filter(([key]) => !getEventsFromMetadata().some((eventKey) => `on${capitalizeFirstLetter(eventKey)}` === key))
        .reduce(
          (acc, [key, value]) => {
            if (getSlotsFromMetadata().includes(key)) {
              acc.slotProps[key] = value;
            } else {
              acc.regularProps[REACT_SPECIFIC_ATTRIBUTES.includes(key) ? key : toKebabCase(key)] = value;
            }
            return acc;
          },
          { regularProps: {}, slotProps: {} }
        );
    };

    // render
    const { regularProps: passedProps, slotProps: actualSlotProps } = getRegularProps();

    const { children, tooltip, ...rest } = passedProps as T & WithWebComponentPropTypes;
    return (
      <CustomTag {...getBooleanProps()} ref={ref} {...rest} class={className} title={tooltip}>
        {Object.entries(actualSlotProps).map(([slotName, slotValue]) => {
          return Children.map(slotValue, (item: ReactElement<any>, index) =>
            cloneElement(item, {
              key: `${slotName}-${index}`,
              slot: slotName
            })
          );
        })}
        {children}
      </CustomTag>
    );
  });

  if (WebComponent) {
    WithWebComponent.displayName = `WithWebComponent(${WebComponent.name})`;
  }

  return WithWebComponent as RefForwardingComponent<Ui5DomRef, T & WithWebComponentPropTypes>;
}
