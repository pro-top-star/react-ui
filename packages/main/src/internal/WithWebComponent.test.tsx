import { mountThemedComponent } from '@shared/tests/utils';
import UI5Button from '@ui5/webcomponents/dist/Button';
import React, { cloneElement, FC } from 'react';
import { spy } from 'sinon';
import { withWebComponent } from './withWebComponent';

describe('withWebComponent', () => {
  test('Unmount Event Handlers correctly after prop update', () => {
    const Button: FC<any> = withWebComponent(UI5Button);
    const callback = spy();
    // eslint-disable-next-line react/jsx-no-bind
    const wrapper = mountThemedComponent(<Button onClick={(...args) => callback(...args)} />);
    const component = wrapper
      .find('ui5-button')
      .first()
      .instance();
    // @ts-ignore
    component.fireEvent('click');
    expect(callback.callCount).toEqual(1);
    wrapper.setProps({
      children: cloneElement(wrapper.prop('children'), { onClick: (...args) => callback(...args) })
    });
    wrapper.update();
    // @ts-ignore
    component.fireEvent('click');
    expect(callback.callCount).toEqual(2);
  });

  test('Bind new event handler', () => {
    const Button: FC<any> = withWebComponent(UI5Button);
    const callback = spy();
    const wrapper = mountThemedComponent(<Button onClick={callback} />);
    const component = wrapper
      .find('ui5-button')
      .first()
      .instance();
    // @ts-ignore
    component.fireEvent('click');
    expect(callback.callCount).toEqual(1);

    const anotherCallback = spy();
    wrapper.setProps({
      children: cloneElement(wrapper.prop('children'), { onClick: anotherCallback })
    });
    wrapper.update();
    // @ts-ignore
    component.fireEvent('click');
    expect(callback.callCount).toEqual(1);
    expect(anotherCallback.callCount).toEqual(1);
  });
});
