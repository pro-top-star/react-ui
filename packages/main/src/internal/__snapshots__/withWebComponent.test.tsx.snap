// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withWebComponent scoping 1`] = `
<DocumentFragment>
  <ui5-button-ui5-wcr
    design="Default"
  >
    Scoping Test
  </ui5-button-ui5-wcr>
</DocumentFragment>
`;

exports[`withWebComponent scoping 2`] = `
<DocumentFragment>
  <ui5-button
    design="Default"
    ui5-button=""
  >
    Scoping Test
  </ui5-button>
</DocumentFragment>
`;
